### PR TITLE
ranger: add imagePreviewSupport option and make previews work out of the box

### PR DESCRIPTION
**Pull request by oxij**
Merge #26452 first or this won't build.

###### Motivation for this change
See commit message.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/misc/ranger/default.nix [hunk 1]
```diff
@@ -1,4 +1,7 @@
-{ stdenv, fetchurl, pythonPackages, w3m, file, less }:
+{ stdenv, fetchurl, pythonPackages, file, less
+, imagePreviewSupport ? true, w3m ? null}:
```
Comment (joachifm) [inline pkgs/applications/misc/ranger/default.nix]: I wonder what the purpose of defaulting w3m to `null` is? As far as I can tell, the expression fails to evaluate if `w3m = null` is in fact passed here (provided `imagePreviewSupport = true`). That'd require user intervention of course, but still seems odd to default to an invalid combination of parameter values.

**Comment by oxij** [comment]
The purpose is to show that `w3m` is in fact optional if you disable
image previews. A lot of packages do the same. I was too lazy to add an
assert because the build time is tiny anyway.

Since there seems to be no standard which "pattern" to use

    { optionSupport ? ..., package ? null }
    assert optionSupport -> package != null;
    ... and be sure to not use the package when optionSupport is false ..
    and then evaluate as .override { optionSupport = false; }
    ... or don't ...
    and then evaluate as .override { optionSupport = false; package = null; }

    This pattern requires the least amount of knowledge on the user
    part. You callPackages without some of the optional packages and it
    will work. I prefer this.

or

    { optionSupport ? ..., package }
    and then
    ... and be sure to not use the package when optionSupport is false ..
    and then evaluate as .override { optionSupport = false; }
    ... or don't ...
    and then evaluate as .override { optionSupport = false; package = null; }

    Okay in nixpkgs context, but can't be evaluated with callPackages
    without some of the optional packages.

or

    { package ? null }
    ... and depend on package being null when the optionSupport should be disabled ...
    and then evaluate as .override { package = null; }

    This works like the first pattern, but can't be overriden by
    `nixpkgs.config`, which I love and use extensively.

or

    { package }
    ... and depend on package being null when the optionSupport should be disabled ...
    and then evaluate as .override { package = null; }

    Many people prefer this because it the least amount of typing in the
    package expression, but I hate it. `buildInputs` ignoring nulls is a
    hack that should be discouraged, IMO.

I do one of the first two whichever it is I feel like ATM. I can either
add an assert, or remove the null. Whichever you prefer, I don't really
care in this case.

**Comment by joachifm** [comment]
Expressing optional features is a bit clunky atm, to put it midlly, and have poor discoverability, but that's a topic I'm not too keen on rehashing, so if you prefer using this pattern that's fine by me. I just think the defaults are misleading here, by suggesting a combination of param values that causes eval error.

**Comment by oxij** [comment]
I think current version is ok to merge.

**Comment by joachifm** [comment]
Thank you

**Pull request merged**